fw/mmap_buffer.c

```diff
@@ -0,0 +1,329 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include "mmap_buffer.h"
+#include "lib/str.h"
+#include <linux/types.h>
+#include <linux/log2.h>
+#include <linux/cpumask.h>
+#include <linux/fs.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <asm/page.h>
+
+/*
+ * We can't pass TfwMmapBufferHolder pointer to the file operations handlers.
+ * Let's store these pointers, and find them by filenames in the open handler.
+ */
+#define MAX_HOLDERS 4
+static TfwMmapBufferHolder *holders[MAX_HOLDERS];
+static int holders_cnt;
+
+static int dev_file_open(struct inode *ino, struct file *filp);
+static int dev_file_close(struct inode *ino, struct file *filp);
+static int dev_file_mmap(struct file *filp, struct vm_area_struct *vma);
+static void dev_file_vm_close(struct vm_area_struct *vma);
+
+static const struct file_operations dev_fops = {
+	.open  = dev_file_open,
+	.release  = dev_file_close,
+	.mmap  = dev_file_mmap,
+};
+
+static const struct vm_operations_struct dev_vm_ops = {
+	.close = dev_file_vm_close
+};
+
+static int
+dev_file_open(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder;
+	int i;
+
+	for (i = 0; i < holders_cnt; ++i) {
+		if (strcmp(holders[i]->dev_name,
```

Comment:
It seems that there can be already freed memory access here. We never clear `holders` array. So after calling `tfw_mmap_buffer_free` here we can access already freed memory. May be we should add appropriate unit test.

Reply:
Before holder freeing we destroy the device. I think it's impossible, that this handler can be called after device destroying.

Reply:
Currently we destroy device on each restarting Tempesta FW. see `tfw_access_log_start`. @krizhanovsky Should we use `tfw_runstate_is_reconfig` in `tfw_access_log_start` to immediately return from `tfw_access_log_start` on reconfiguration or we should implement full deinitialization here?

Reply:
@EvgeniiMekhanik I'm not sure if I understood your question, but it seems you are who can answer it

Reply:
Now config reload doesn't  cause mmap_buffer recreation. 
```
	if (!(access_log_type & ACCESS_LOG_MMAP) || mmap_buffer)
		return 0;
```